planner/cascades/transformation_rules.go

```diff
@@ -634,3 +637,55 @@ func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.
 	}
 	return finalGroupExprs, true, false, nil
 }
+
+// PushTopNDownProjection pushes TopN to Projection.
+type PushTopNDownProjection struct {
+	baseRule
+}
+
+// NewRulePushTopNDownProjection creates a new Transformation PushTopNDownProjection.
+// The pattern of this rule is `TopN->Projection->X` to `Projection->TopN->X`.
+func NewRulePushTopNDownProjection() Transformation {
+	rule := &PushTopNDownProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the TopN through Projection.
+func (r *PushTopNDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	proj := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].GetExpr().Children[0]
+
+	newTopN := plannercore.LogicalTopN{
+		Offset: topN.Offset,
+		Count:  topN.Count,
+	}.Init(topN.SCtx(), topN.SelectBlockOffset())
+
+	newTopN.ByItems = make([]*plannercore.ByItems, 0, len(topN.ByItems))
+	for _, by := range topN.ByItems {
+		newTopN.ByItems = append(newTopN.ByItems, &plannercore.ByItems{
+			Expr: expression.ColumnSubstitute(by.Expr, old.Children[0].Group.Prop.Schema, proj.Exprs),
```

Comment:
Can we push down the topn regardless of the exprs? If the expr has side effects, the result maybe incorrect. For example, t has 5 rows, in TiDB:
```
mysql> set @i = 0;
Query OK, 0 rows affected (0.00 sec)

mysql>  select a, b from (select @i as a, @i := @i+1 as b from t) t order by a desc limit 1;
+------+------+
| a    | b    |
+------+------+
| 0    | 1    |
+------+------+
1 row in set (0.00 sec)
```

While in MySQL,
```
mysql> set @i = 0;
Query OK, 0 rows affected (0.00 sec)

mysql> select a, b from (select @i as a, @i := @i+1 as b from t) t order by a desc limit 1;
+------+------+
| a    | b    |
+------+------+
|    4 |    5 |
+------+------+
1 row in set, 1 warning (0.01 sec)
```
```

Reply:
So I can just use the old TopN's items without call `expression.ColumnSubstitute` ?

Reply:
No, I mean we cannot push down every expresison. You can check https://github.com/pingcap/tidb/blob/master/planner/core/rule_predicate_push_down.go#L365 to filter these kind of expressions out.

Reply:
Updated and PTAL. Thanks